### PR TITLE
190608

### PR DESCRIPTION
**Pull request by SunilSimha**
It contains disparate mini-projects including:

- CIGALE wrapper
- EAZY wrapper
- KCWI datacube handling tools.
- Extinction correction module.

and other code written to facilitate the 190608 host and fg papers. Some code for photometry to flux conversion has kinks on this branch that have been fixed recently in `heintz2020`. That being said, it should not affect any analysis performed in the host or fg papers because the bugs pertain to VISTA photometry in particular.

### CODE REVIEW

File: frb/analysis/kcwi.py [hunk 1]
```diff
@@ -0,0 +1,470 @@
...
+
+from astropy.io import fits
+from astropy.table import Table
+from astropy.stats import sigma_clipped_stats
+from astropy.coordinates import SkyCoord
+from astropy.wcs import WCS, utils as wcsutils
+from astropy import units as u
+from scipy.interpolate import interp1d
+
+try:
```
[the remark this answers is not in the document]
Comment (JayChittidi) [reply frb/analysis/kcwi.py:16]: Aren't both scipy and astropy already in the Install docs?

File: frb/galaxies/cigale.py [hunk 2]
```diff
@@ -335,3 +340,20 @@ def run(photometry_table, zcol, data_file="cigale_in.fits", config_file="pcigale
             #import pdb; pdb.set_trace()
             
     return
+
+def host_run(photom, host, cigale_file=None):
+    cigale_tbl = photom.copy()
```
[the remark this answers is not in the document]
Comment (SunilSimha) [reply frb/galaxies/cigale.py:345]: Done.

File: frb/halos.py [hunk 3]
```diff
@@ -424,30 +432,52 @@ def stellarmass_from_halomass(log_Mhalo):
...
     return log_mstar
 
 
-def halomass_from_stellarmass(log_mstar):
-    """ Halo mass from Stellar mass (Moster+2010)
+def halomass_from_stellarmass(log_mstar,z=0):
+    """ Halo mass from Stellar mass (Moster+2013)
 
     Args:
         log_mstar (float): log10 of the galaxy stellar mass (solar masses)
```
[the remark this answers is not in the document]
Comment (SunilSimha) [reply frb/halos.py]: Done.

File: frb/igm.py [hunk 4]
```diff
@@ -208,96 +255,111 @@ def average_DM(z, cosmo=None, cumul=False, neval=10000, mu=1.3):
...
-        z (float):
-            Redshift for the evaluation
-        fb (float, optional):
-            Fraction of the cosmic baryons in a given halo
-            1 means halos have all of their alloted baryons to r200
-        rmax (float, optional):
-        neval (int, optional):
-            Number of redshift evaluations
-
+        z: float
```
[the remark this answers is not in the document]
Comment (SunilSimha) [reply frb/igm.py]: Resolved

File: frb/igm.py [hunk 5]
```diff
@@ -153,8 +153,72 @@ def z_from_DM(DM, cosmo=None, coord=None, corr_nuisance=True):
...
+  if not return_rho:
+    return f_diffuse
+  else:
+    return f_diffuse, rho_b*f_diffuse*(1+z)**3
+
+def ne_cosmic(z, cosmo = Planck15, mu = 4./3):
+  """
+  Calculate the average cosmic electron
+  number density as a function of redshift.
+  Args:
```
[the remark this answers is not in the document]
Comment (SunilSimha) [reply frb/igm.py:187]: Resolved.

File: frb/igm.py [hunk 6]
```diff
@@ -153,8 +153,72 @@ def z_from_DM(DM, cosmo=None, coord=None, corr_nuisance=True):
...
 
+def f_diffuse(z, cosmo = Planck15, return_rho = False):
+  """
+  Calculate the cosmic fraction of baryons
+  in diffuse gas phase based on our empirical
+  knowledge of baryon distributions and their 
+  ionization state.
 
-def average_DM(z, cosmo=None, cumul=False, neval=10000, mu=1.3):
+  Args:
```
[the remark this answers is not in the document]
Comment (SunilSimha) [reply frb/igm.py:157]: Resolved.

File: frb/analysis/kcwi.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (SunilSimha) [reply frb/analysis/kcwi.py:16]: Yep. Scipy and Astropy already exist.

File: frb/analysis/kcwi.py [hunk 7]
```diff
@@ -0,0 +1,470 @@
...
+    import sep
+except ImportError:
+    raise ImportError("Requirement unmet: sep. Run `pip install sep`")
+
+try:
```
[the remark this answers is not in the document]
Comment (SunilSimha) [reply frb/analysis/kcwi.py:21]: Done.

**Pull request merged**